### ISSUE TITLE
`log_softmax` could be `2**102` to `2**970` times more accurate

### ISSUE THREAD

**Issue by JasonGross** [tensorflow/tensorflow#62400]
### Issue type

Feature Request

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

unknown 2.12.0

### Custom code

No

### OS platform and distribution

Linux 5.15.0-67-generic #74~20.04.1-Ubuntu SMP Wed Feb 22 14:52:34 UTC 2023 x86_64 GNU/Linux

### Python version

3.11.4

### Current behavior?

As I understand it, TensorFlow implements `log_softmax(x)` as `x - max(x) - log(sum(exp(x - max(x))))`. However, when the largest value is much larger than the rest of the values (about 16 larger for float32, about 36 larger for float64), `log_softmax` returns `0` at the maximum value, when it could give a much more precise answer by using `log1p` and zeroing out the maximum cell (guaranteed to be 1) before summing.

Might help with https://github.com/tensorflow/tensorflow/issues/48816, https://github.com/tensorflow/tensorflow/issues/12002

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
eps = tf.constant(tf.keras.backend.epsilon(), dtype=tf.float32)
print(-tf.nn.log_softmax(tf.constant([1 - tf.math.log(2 * eps).numpy(), 0], dtype=tf.float32))) # tf.Tensor([1.1920928e-07 1.6424950e+01], shape=(2,), dtype=float32)
print(-tf.nn.log_softmax(tf.constant([1 - tf.math.log(eps).numpy(), 0], dtype=tf.float32))) # tf.Tensor([-0.       17.118095], shape=(2,), dtype=float32)


```python
import tensorflow as tf
import numpy as np

def log_softmax_alt(x):
    maxi = tf.argmax(x)
    xoffset = x - x[maxi]
    xoffsetexp = tf.exp(xoffset)
    # xoffsetexp[maxi] is currently about 1
    xoffsetexp = tf.tensor_scatter_nd_update(xoffsetexp, [[maxi]], [0])
    xoffsetexp_sum_m1 = tf.reduce_sum(xoffsetexp)
    return xoffset - tf.math.log1p(xoffsetexp_sum_m1)


for ty in (tf.float32, tf.float64):
    smallest_log_softmax, smallest_log_softmax_alt, smallest_log_softmax_val, smallest_log_softmax_alt_val = 0, 0, 0, 0
    # work around https://github.com/tensorflow/tensorflow/issues/62399
    for i in range(int(1 - np.log2(tf.experimental.numpy.finfo(ty).smallest_subnormal))):
        values = tf.constant([1 + i, 0], dtype=ty)
        log_softmax_values = tf.nn.log_softmax(values)
        log_softmax_values_alt = log_softmax_alt(values)
        if log_softmax_values[0] != 0: smallest_log_softmax, smallest_log_softmax_val = i, log_softmax_values[0]
        if log_softmax_values_alt[0] != 0: smallest_log_softmax_alt, smallest_log_softmax_alt_val = i, log_softmax_values_alt[0]
        if log_softmax_values[0] == 0 and log_softmax_values_alt[0] == 0: break
    print(f"For {ty}, diff in supported input accuracy is 2**-({smallest_log_softmax_alt} - {smallest_log_softmax}) = 2**-{smallest_log_softmax_alt - smallest_log_softmax}; diff in output accuracy is np.log2({-smallest_log_softmax_val}) - np.log2({-smallest_log_softmax_alt_val}) = {np.log2(-smallest_log_softmax_val) - np.log2(-smallest_log_softmax_alt_val)}")
```
gives
```
For <dtype: 'float32'>, diff in supported input accuracy is 2**-(86 - 15) = 2**-71; diff in output accuracy is np.log2(1.1920928244535389e-07) - np.log2(1.6458114537543937e-38) = 102.51446533203125
For <dtype: 'float64'>, diff in supported input accuracy is 2**-(707 - 35) = 2**-672; diff in output accuracy is np.log2(2.2204460492503128e-16) - np.log2(3.3075530036384083e-308) = 969.428088949386
```
```


### Relevant log output

_No response_

**Comment by sushreebarsa** [comment]
@sachinprasadhs  I was able to replicate the issue reported [here](https://colab.research.google.com/gist/sushreebarsa/00223aae8b2afeab0da98122da85a040/62400.ipynb). Please find the attached gist. Thank you!

**Comment by sanskarmodi8** [comment]
If no one is working on this, I was thinking to work on this and create a Pull Request but I am unable to find the actual implementation of log_softmax function in Tensorflow. Can someone guide me about it?

**Comment by JasonGross** [comment]
I am guessing it's
https://github.com/tensorflow/tensorflow/blob/c1c9184f35353b48909b1cf633185231b6a59c7c/tensorflow/lite/kernels/internal/reference/integer_ops/log_softmax.h#L25-L108
and https://github.com/tensorflow/tensorflow/blob/c1c9184f35353b48909b1cf633185231b6a59c7c/tensorflow/compiler/tf2tensorrt/convert/ops/log_softmax.cc#L63-L83
and
https://github.com/tensorflow/tensorflow/blob/c1c9184f35353b48909b1cf633185231b6a59c7c/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L3819-L3839
and maybe a bunch of other files returned by https://github.com/search?q=repo%3Atensorflow%2Ftensorflow+logsoftmax+language%3AC%2B%2B&type=code

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by JasonGross** [comment]
The issue is not exactly hard to check.  If you cared about fixing issues, wouldn't you just check it yourself, rather than offloading the burden onto issue reporters?  Did anyone explicitly work on fixing this?

**Comment by Kayyuri** [comment]
Hi @JasonGross,
Thank you so much for filing this Feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
